documentation/wiki/product/decisions/adr/2025-07-23-file-upload-virus-scanning.md

```diff
@@ -0,0 +1,148 @@
+
+To ensure the security and integrity of the platform, we need to implement virus scanning for all uploaded files before they are stored and made available for download.
+
+_What is the best architectural approach to implement virus scanning for file uploads?_
+
+## Options Considered
+
+* **In-App Server ClamAV Scanning** - Run ClamAV daemon directly in the Flask application container
+* **Asynchronous S3 Event-Driven Scanning** - Upload to staging bucket, scan via S3 events, move to final bucket
+* **Third-party SaaS Solution** - Use services like VirusTotal API or AWS GuardDuty Malware Protection
+
+## Decision Recommendation
+
+Recommend **In-App Server ClamAV Scanning** by running the ClamAV daemon directly within our existing Flask application containers.
+
+### Architecture Components
+
+1. **ClamAV Daemon** - Runs as a background process in each Flask app container
+2. **Enhanced Upload Endpoint** - Modified `/applications/{id}/attachments` endpoint with virus scanning
+3. **Python ClamAV Client** - Uses `clamd` library to communicate with local ClamAV daemon
+4. **Definition Updates** - Scheduled `freshclam` updates within containers every 6 hours
+5. **Error Handling** - Return immediate 422 errors for infected files
+
+### Implementation Approach
+
+```python
+# Add to app initialization
+import clamd
+
+# Initialize ClamAV connection (once at app startup)
+clam = clamd.ClamdUnixSocket()
+
+# Modified create_application_attachment service
+def create_application_attachment(db_session, application_id, user, form_and_files_data):
+    file_attachment = form_and_files_data.get("file_attachment")
+    
+    # Pre-scan validation
+    if file_attachment.content_length > MAX_SCAN_SIZE:
+        raise_flask_error(422, "File too large for virus scanning")
+    
+    # Perform virus scan using local ClamAV daemon
+    scan_result = clam.instream(file_attachment.stream)
+    
+    if scan_result['stream'][0] == 'FOUND':
+        threat_name = scan_result['stream'][1]
+        logger.warning(f"Malware detected in upload: {threat_name}")
+        raise_flask_error(422, f"File rejected: malware detected ({threat_name})")
+    
+    if scan_result['stream'][0] == 'ERROR':
+        logger.error(f"Virus scan failed: {scan_result['stream'][1]}")
+        raise_flask_error(500, "File upload temporarily unavailable")
+    
+    # Proceed with S3 upload only if clean
+    s3_key = file_util.open_stream(file_attachment)
+    # ... rest of attachment creation logic
+```
+
+### Docker Configuration
+
+```dockerfile
+# Add to existing Dockerfile
+RUN apt-get update && apt-get install -y \
+    clamav \
+    clamav-daemon \
+    clamav-freshclam \
+    && rm -rf /var/lib/apt/lists/*
+
+# Update virus definitions
+RUN freshclam
+
+# Start ClamAV daemon
+RUN mkdir -p /var/run/clamav && \
+    chown clamav:clamav /var/run/clamav
+    
+# Add to container startup script
+CMD ["sh", "-c", "clamd --config-file=/etc/clamav/clamd.conf & python app.py"]
+```
```

Comment:
Is there a way we could only have clam AV running when we run as the API (some env var we could set?). We wouldn't need virus scanning on backend tasks since it doesn't interact with live user data and adding more RAM to support something we aren't going to call seems wasteful.

Alternatively, is there any way to run clamav as a sidecar in its own container so it can get its own resources, and then maybe when we startup a task we say whether we want that sidecar (if that's possible?).